src/history.coffee

```diff
@@ -118,13 +133,18 @@ class History
         hasSeenChanges = true
     null
 
-  popChanges: (fromStack, toStack, checkpointIndex) ->
+  popChanges: (fromStack, toStack, checkpointIndex, lookBack) ->
     changes = []
-    snapshot = fromStack[checkpointIndex].snapshot
-    for entry in fromStack.splice(checkpointIndex) by -1
+    snapshotBefore = null
+    snapshotAfter = fromStack[checkpointIndex].snapshot
+    splicedEntries = fromStack.splice(checkpointIndex)
+    for entry in splicedEntries by -1
       toStack?.push(entry)
-      changes.push(entry) unless entry instanceof Checkpoint
-    {changes, snapshot}
+      if entry instanceof Checkpoint
```

Comment:
In the case of `redo`, we want `snapshotBefore` to be the snapshot taken *right* before the change, such that it will then become `snapshotAfter` if we `undo` again.

So if you had manually created some checkpoints via `::createCheckpoint()`, the redo stack might look like this:

```
... [checkpoint] [CHANGE] [CHANGE] [checkpoint] [checkpoint] [checkpoint] [checkpoint]
         ^- snapshotAfter                ^- snapshotBefore
```